### DISCUSSION TITLE
In-Memory databases with embedded server

### DISCUSSION THREAD

**Discussion by toochevere** [ravendb/ravendb#15372] [Q&A] [answered]
I need some guidance about how to use in-memory only databases with the embedded server.  

The use case is to speed up an extensive set of tests for a library that uses Raven.  When starting the embedded server, it is clear how to set the data directory and logs directory, but it seems no option is present for keeping databases purely in memory.   Older versions of Raven offered an in-memory option that no longer seems to be present.  

Looking at the code, it would seem that the only way to control this is via `settings.json`?  Is there a better way?

**Comment by ayende** [comment, marked as answer]
There is the `Raven.TestDriver` that does that for you.
Is there a reason you use the embedded server instead of the builtin facilities?

You can define `RunInMemory` at the server or per database level, to control this.

**Reply by toochevere** [reply]
Thank you for the answer.   A few questions:

Regarding using the test driver
* I can use the test driver for tests that employ the testing framework.  But I also have functional tests that are going to be driven by a another process.  In this case, I was looking at simply enabling the in-memory option.
* I still didn't see an "in-memory" option on the test driver.  I suppose that it keeps databases in memory by default?  I had observed that it still creates folders for the database but after your answer I looked a little closer and realized that nothing significant was being written there.  

In order to set this at the database level, do I just set the option via a `PutDatabaseSettingsOperation`?   
```
ravenDocumentStore.Maintenance.Send(new PutDatabaseSettingsOperation(
    dbName,
    new Dictionary<string, string> {
        { "Core.RunInMemory" , "true" }
    }));
```

EDIT:  I tried setting this with both "Core.RunInMemory" and "RunInMemory" and observed that in both cases files were still created on disk.  

EDIT2:  Looking at [how the test driver does it](https://github.com/ravendb/ravendb/blob/a40b34616332911a55e3c2e28209a6aab7999b44/src/Raven.TestDriver/RavenTestDriver.cs#L274), this is sufficient for my purposes.  It would be nice to understand how to do this on a database level, but not absolutely required. 
```
options.CommandLineArgs.Add("--RunInMemory=true")
```

**Reply by garayx** [reply]
You can create an inmemory database like this:

```
            DocumentStore.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord("test")
            {
                Settings = new Dictionary<string, string>() { { "Core.RunInMemory", "true" } }
            }));
```
